src/app/clusters/occupancy-sensor-server/occupancy-sensor-server.cpp

```diff
@@ -16,14 +16,144 @@
  */
 
 #include "occupancy-sensor-server.h"
+#include "occupancy-hal.h"
 
-#include <app-common/zap-generated/attributes/Accessors.h>
+#include <app/AttributeAccessInterfaceRegistry.h>
+#include <app/EventLogging.h>
+#include <app/data-model/Encode.h>
+#include <app/util/attribute-storage.h>
+#include <lib/core/CHIPError.h>
 
-#include "occupancy-hal.h"
+using chip::Protocols::InteractionModel::Status;
+
+namespace chip {
+namespace app {
+namespace Clusters {
+namespace OccupancySensing {
+
+Structs::HoldTimeLimitsStruct::Type * HoldTimeLimitsManager::HoldTimeLimits::GetHoldTimeLimitsStruct(EndpointId endpoint)
+{
+    size_t endpointIndex                                       = 0;
+    Structs::HoldTimeLimitsStruct::Type * holdTimeLimitsStruct = nullptr;
+    CHIP_ERROR status                                          = FindHoldTimeLimitsIndex(endpoint, endpointIndex);
+    if (CHIP_NO_ERROR == status)
+    {
+        holdTimeLimitsStruct = &mHoldTimeLimitsStructs[endpointIndex];
+    }
+    return holdTimeLimitsStruct;
+}
+
+CHIP_ERROR
+HoldTimeLimitsManager::HoldTimeLimits::SetHoldTimeLimitsStruct(EndpointId endpoint,
+                                                               Structs::HoldTimeLimitsStruct::Type & holdTimeLimitsStruct)
+{
+    VerifyOrReturnError(kInvalidEndpointId != endpoint, CHIP_ERROR_INVALID_ARGUMENT);
+
+    size_t endpointIndex = 0;
+    ReturnErrorOnFailure(FindHoldTimeLimitsIndex(endpoint, endpointIndex));
+
+    mHoldTimeLimitsStructs[endpointIndex] = holdTimeLimitsStruct;
+
+    return CHIP_NO_ERROR;
+}
+
+/// @brief Returns the index of the HoldTimeLimits associated to an endpoint
+/// @param[in] endpoint target endpoint
+/// @param[out] endpointIndex index of the corresponding HoldTimeLimits for an endpoint
+/// @return CHIP_NO_ERROR or CHIP_ERROR_NOT_FOUND, CHIP_ERROR_INVALID_ARGUMENT if invalid endpoint
+CHIP_ERROR HoldTimeLimitsManager::HoldTimeLimits::FindHoldTimeLimitsIndex(EndpointId endpoint, size_t & endpointIndex)
+{
+    VerifyOrReturnError(kInvalidEndpointId != endpoint, CHIP_ERROR_INVALID_ARGUMENT);
+
+    uint16_t index = emberAfGetClusterServerEndpointIndex(endpoint, OccupancySensing::Id,
+                                                          MATTER_DM_OCCUPANCY_SENSING_CLUSTER_SERVER_ENDPOINT_COUNT);
+
+    if (index < ArraySize(mHoldTimeLimitsStructs))
+    {
+        endpointIndex = index;
+        return CHIP_NO_ERROR;
+    }
+    return CHIP_ERROR_NOT_FOUND;
+}
+
+HoldTimeLimitsManager HoldTimeLimitsManager::mInstance;
+
+HoldTimeLimitsManager & HoldTimeLimitsManager::Instance()
+{
+    return mInstance;
+}
+
+CHIP_ERROR HoldTimeLimitsManager::Init()
+{
+    // Prevents re-initializing
+    VerifyOrReturnError(!mIsInitialized, CHIP_ERROR_INCORRECT_STATE);
+
+    for (size_t i = 0; i <= kOccupancySensingServerMaxEndpointCount; i++)
+    {
+        if (emberAfContainsServer(EndpointId(i), OccupancySensing::Id))
```

Comment:
This is not going to work right for, say, a bridge that needs to add these endpoints dynamically.

One possible place to initialize the entry for an enpoint id is when this cluster is actually initialized on that endpoint.   See `emberAfOccupancySensingClusterServerInitCallback` below.

That's for initialization the SDK itself can do.  For initialization the application does, they would need to implement `emberAfOccupancySensorInitCallback` (note no "Server" in the name), assuming you stick to the ember-callbacks method of initializing things.

Reply:
Changed in new code, do the initialization in application code, please check